File: android/expoview/src/main/java/host/exp/exponent/experience/DetachActivity.java
// Copyright 2015-present 650 Industries. All rights reserved.

package host.exp.exponent.experience;

import android.content.Intent;
import android.os.Bundle;

import com.facebook.react.ReactPackage;
import com.facebook.react.ReactRootView;

import java.util.List;

import org.unimodules.adapters.react.ReactModuleRegistryProvider;
import org.unimodules.core.interfaces.Package;
import org.unimodules.core.interfaces.SingletonModule;

import expo.modules.splashscreen.SplashScreenImageResizeMode;
import expo.modules.splashscreen.singletons.SplashScreen;
import host.exp.exponent.Constants;
import host.exp.expoview.ExpoViewBuildConfig;
import versioned.host.exp.exponent.ExponentPackageDelegate;
import versioned.host.exp.exponent.modules.universal.ExpoModuleRegistryAdapter;

public abstract class DetachActivity extends ExperienceActivity implements ExponentPackageDelegate {

  // Override me!
  public abstract String publishedUrl();
  public abstract String developmentUrl();
  public abstract List<ReactPackage> reactPackages();
  public abstract List<Package> expoPackages();
  public abstract boolean isDebug();

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    ExpoViewBuildConfig.DEBUG = isDebug();
    Constants.INITIAL_URL = isDebug() ? developmentUrl() : publishedUrl();
    mManifestUrl = Constants.INITIAL_URL;

    if (getIntent().getData() != null) {
      mIntentUri = getIntent().getData().toString();
    }

    super.onCreate(savedInstanceState);

    SplashScreen.show(this, Constants.SPLASH_SCREEN_IMAGE_RESIZE_MODE, ReactRootView.class, true);

    mKernel.handleIntent(this, getIntent());
  }

  @Override
  public void onNewIntent(Intent intent) {
    super.onNewIntent(intent);
    mKernel.handleIntent(this, intent);
  }

  // TODO: eric: make Constants.INITIAL_URI reliable so we can get rid of this
  @Override
  public void shouldCheckOptions() {
    if (mManifestUrl != null && mKernel.hasOptionsForManifestUrl(mManifestUrl)) {
      handleOptions(mKernel.popOptionsForManifestUrl(mManifestUrl));
    } else if (isDebug() && mKernel.hasOptionsForManifestUrl(publishedUrl())) {
      // also check publishedUrl since this can get set before Constants.INITIAL_URL is set to developmentUrl
      handleOptions(mKernel.popOptionsForManifestUrl(publishedUrl()));
    }
  }

  @Override
  public ExponentPackageDelegate getExponentPackageDelegate() {
    return this;
  }

  @Override
  public ExpoModuleRegistryAdapter getScopedModuleRegistryAdapterForPackages(List<Package> packages, List<SingletonModule> singletonModules) {
    return new DetachedModuleRegistryAdapter(new ReactModuleRegistryProvider(packages, singletonModules));
  }
}

File: android/expoview/src/main/java/host/exp/exponent/experience/DetachActivity.kt
// Copyright 2015-present 650 Industries. All rights reserved.
package host.exp.exponent.experience

import android.content.Intent
import android.os.Bundle
import com.facebook.react.ReactPackage
import com.facebook.react.ReactRootView
import expo.modules.adapters.react.ReactModuleRegistryProvider
import expo.modules.core.interfaces.Package
import expo.modules.core.interfaces.SingletonModule
import expo.modules.splashscreen.singletons.SplashScreen
import host.exp.exponent.Constants
import host.exp.expoview.ExpoViewBuildConfig
import versioned.host.exp.exponent.ExponentPackageDelegate
import versioned.host.exp.exponent.modules.universal.ExpoModuleRegistryAdapter

abstract class DetachActivity : ExperienceActivity(), ExponentPackageDelegate {
  // Override me!
  abstract fun publishedUrl(): String?
  abstract fun developmentUrl(): String?
  abstract override fun reactPackages(): List<ReactPackage>?
  abstract override fun expoPackages(): List<Package>?
  abstract val isDebug: Boolean

  override fun onCreate(savedInstanceState: Bundle?) {
    ExpoViewBuildConfig.DEBUG = isDebug
    Constants.INITIAL_URL = if (isDebug) developmentUrl() else publishedUrl()
    manifestUrl = Constants.INITIAL_URL

    if (intent.data != null) {
      intentUri = intent.data.toString()
    }

    super.onCreate(savedInstanceState)

    SplashScreen.show(this, Constants.SPLASH_SCREEN_IMAGE_RESIZE_MODE, ReactRootView::class.java, true)

    kernel.handleIntent(this, intent)
  }

  override fun onNewIntent(intent: Intent) {
    super.onNewIntent(intent)
    kernel.handleIntent(this, intent)
  }

  // TODO: eric: make Constants.INITIAL_URI reliable so we can get rid of this
  override fun shouldCheckOptions() {
    if (manifestUrl != null && kernel.hasOptionsForManifestUrl(manifestUrl)) {
      handleOptions(kernel.popOptionsForManifestUrl(manifestUrl)!!)
    } else if (isDebug && kernel.hasOptionsForManifestUrl(publishedUrl())) {
      // also check publishedUrl since this can get set before Constants.INITIAL_URL is set to developmentUrl
      handleOptions(kernel.popOptionsForManifestUrl(publishedUrl())!!)
    }
  }

  override val exponentPackageDelegate: ExponentPackageDelegate
    get() = this

  override fun getScopedModuleRegistryAdapterForPackages(
    packages: List<Package>,
    singletonModules: List<SingletonModule>
  ): ExpoModuleRegistryAdapter {
    return DetachedModuleRegistryAdapter(ReactModuleRegistryProvider(packages, singletonModules))
  }
}
